### PR TITLE
opal/mca/threads/qthreads: Fix #8036

### LINKED ISSUE #8036: Argobots and Qthread configure detection need fixing/improvements
Working with current master (`g560ebc5780`), I am trying to compile Open MPI with either Argobots or Qthread support and noticed the following:

1) There had been a patch to improve the detection of Argobots in `configure` (https://github.com/open-mpi/ompi/pull/7675/files) but the current state does not seem to work. I am passing `--with-argobots=$HOME/opt/argobots-1.0/` to configure but all I am getting is the following message in `config.log`:
```
configure:5285: +++ Configuring MCA framework threads
configure:153773: checking for no configure components in framework threads
configure:153775: result: 
configure:153777: checking for m4 configure components in framework threads
configure:153779: result: argobots, pthreads, qthreads
configure:5300: --- MCA component threads:argobots (m4 configuration macro, priority 30)
configure:153867: checking for MCA component threads:argobots compile mode
configure:153870: result: static 
configure:156050: checking if MCA component threads:argobots can compile
configure:156052: result: no
```
Any ideas how to get more information on what is going wrong? I don't see a failed compilation attempt anywhere in the logs...

2) There is no way to provide `configure` with a path to a custom qthread installation. Something along the lines of `--with-qthreads=/path/to/qthreads/` should be provided.

### PR DESCRIPTION
**Pull request by shintaro-iwasaki**
This PR fixes #8036. Specifically, this PR provides the following:
1. Add `--with-qthreads` (the same as Argobots, which has been already added by #7675. I'd like to thank @hppritcha )
2. Fix Qthreads compilation issues reported by @devreal in #8036
3. Implement several not-implemented features for MCA/threads/Qthreads backend.

This PR has been developed together with the Qthreads developers (@olivier-snl and @janciesko). I checked it with the latest Qthreads (you need the latest master branch).

Note that this PR does not affect the behavior of Open MPI by default; this change is effective only when one enables Qthreads.

### CODE REVIEW

**Review by devreal — CHANGES_REQUESTED** [review]
Thanks for working on that :+1: Some small comments inline

**Comment by jsquyres** [comment]
ok to test

File: opal/mca/threads/qthreads/Makefile.am
Comment (jsquyres) [inline opal/mca/threads/qthreads/Makefile.am:36]: Nit: no need to commit commented-out code.

File: opal/mca/threads/qthreads/Makefile.am
Comment (jsquyres) [inline opal/mca/threads/qthreads/Makefile.am:58]: Nit: blank lines at the end of the file.

File: opal/mca/threads/qthreads/configure.m4
Comment (jsquyres) [inline opal/mca/threads/qthreads/configure.m4:75]: I believe that AC_CONFIG_FILES and AC_SUBST are always executed, regardless of the result of the AS_IF.  I.e., these should probably be outside the AS_IF so as not to imply that they are actually conditionally executed.

File: opal/mca/threads/qthreads/configure.m4
Comment (jsquyres) [inline opal/mca/threads/qthreads/configure.m4:129]: Same as above.

I didn't check the `threads` framework, but it's a little odd that a component is responsible for AC_SUBST'ing `TPKG_*` and `THREAD_*` values.  Specifically:

1. Are both `TPKG_*` and `THREAD_*` used by the framework?
1. If so, why isn't the framework doing the `AC_SUBST` (vs. the components)?

File: opal/mca/threads/qthreads/threads_qthreads_module.c
Comment (jsquyres) [inline opal/mca/threads/qthreads/threads_qthreads_module.c:34]: Nit: is there a reason to use `int` here instead of `bool`?

File: opal/mca/threads/qthreads/threads_qthreads_mutex.c
Comment (jsquyres) [inline opal/mca/threads/qthreads/threads_qthreads_mutex.c:98]: Nit: please use `{}` around all blocks, even 1-line blocks.

File: opal/mca/threads/qthreads/threads_qthreads_module.c
Comment (shintaro-iwasaki) [reply opal/mca/threads/qthreads/threads_qthreads_module.c]: Thank you for reviewing. I fixed it.

File: opal/mca/threads/qthreads/threads_qthreads_module.c
Comment (shintaro-iwasaki) [reply opal/mca/threads/qthreads/threads_qthreads_module.c:88]: You are right. I removed it. This module relies on lazy initialization, so I did not need to initialize Qthreads here.

File: opal/mca/threads/qthreads/threads_qthreads_mutex.c
Comment (shintaro-iwasaki) [reply opal/mca/threads/qthreads/threads_qthreads_mutex.c:96]: This lock (or any memory barrier that can order accesses to `m_signaled`) is necessary.
`m_signaled` can be updated by another thread that runs on another worker (=Pthreads), but in my understanding, the C standard does not guarantee when a reader can see `m_signaled` updated by a writer without atomic operations.

Since this `waiter` object is allocated in a function stack, it can cause SEGV if the writer accesses this object after `m_signaled` is read by a reader.  It should not happen in `opal_cond_signal()` if all the memory write operations are issued in the order written in the code.

If acquire-release or seq-cst load/store can remove this lock, but it makes the code a bit more complicated.

File: opal/mca/threads/qthreads/threads_qthreads_wait_sync.c
Comment (shintaro-iwasaki) [reply opal/mca/threads/qthreads/threads_qthreads_wait_sync.c:33]: Basically both Qthreads and Argobots "copy" the implementation of Pthreads, so once #8037 is merged, maybe we can declare them in a common header.

This unification needs some work, though.  Specifically, this structure (`ompi_wait_sync_t`) is different https://github.com/open-mpi/ompi/blob/master/opal/mca/threads/pthreads/threads_pthreads_wait_sync.h#L30-L38.  To abstract these operations, we need to change the Pthreads code, which needs more careful examination.

File: opal/mca/threads/qthreads/configure.m4
Comment (shintaro-iwasaki) [reply opal/mca/threads/qthreads/configure.m4:56]: Thanks.  It relies on the latest master, but this change cannot be observed at a C level.  Specifically, the latest Qthreads changes the way of "#include", which is hard to detect programmatically.  Symbol/version check is promising once Qthreads has an official release that includes this update.

File: opal/mca/threads/qthreads/Makefile.am
Comment (shintaro-iwasaki) [reply opal/mca/threads/qthreads/Makefile.am]: Thank you for reviewing.  I removed it.

File: opal/mca/threads/qthreads/Makefile.am
Comment (shintaro-iwasaki) [reply opal/mca/threads/qthreads/Makefile.am]: I removed the extra line.

File: opal/mca/threads/qthreads/configure.m4
Comment (shintaro-iwasaki) [reply opal/mca/threads/qthreads/configure.m4]: I moved them outside `AS_IF()`.

> AC_CONFIG_FILES and AC_SUBST are always executed

Yes they seem so. I did not know that. Thank you!

File: opal/mca/threads/qthreads/configure.m4
Comment (shintaro-iwasaki) [reply opal/mca/threads/qthreads/configure.m4]: I moved `AC_SUBST()` outside `AS_IF()`.

This implementation is basically a copy from #7675.

1. `TPKG_*` is used only in `opal/mca/thread`. `THREAD_` is used in the top `configure.ac` (https://github.com/open-mpi/ompi/blob/master/configure.ac#L1114-L1117).
2. I believe it is to propagate this "thread" information to the top `configure`, but I do not fully understand the modularity/design of Open MPI. @hppritcha would you have any idea about this?

File: opal/mca/threads/qthreads/threads_qthreads_module.c
Comment (shintaro-iwasaki) [reply opal/mca/threads/qthreads/threads_qthreads_module.c]: Thank you. `bool` is used here in the revised version.

File: opal/mca/threads/qthreads/threads_qthreads_mutex.c
Comment (shintaro-iwasaki) [reply opal/mca/threads/qthreads/threads_qthreads_mutex.c:98]: I fixed it.  I checked all of my changes and I believe I fixed all of one-liner `if`/`for`/`while`/....

File: opal/mca/threads/qthreads/threads_qthreads_module.c
Comment (shintaro-iwasaki) [reply opal/mca/threads/qthreads/threads_qthreads_module.c:88]: For this line, I need `opal_threads_ensure_init_qthreads()` since a static internal function, `opal_thread_get_qthreads_self()`, calls a Qthreads function.

File: opal/mca/threads/qthreads/threads_qthreads_mutex.c
Comment (devreal) [reply opal/mca/threads/qthreads/threads_qthreads_mutex.c:96]: Ahh I see what you mean. The problem is that you are issuing a lot of atomic memory operations, esp. if `qthread_yield` is a no-op (i.e., if there are no other qthreads available for execution). In that short snippet, it's four atomic memory ops already. If you have multiple threads (not qthreads) waiting on this conditional variable these locks will be quite contended. Plus, `opal_mutex_unlock` calls `qthread_yield` again (which I am also not convinced is necessary).

Anyway, what is the `lock` argument in this function good for? It doesn't seem to protect anything so it can be re-taken right before exiting the function (the `cond` has its own lock). If you make `waiter.m_signaled` `volatile` (to prevent the compiler from optimizing away the loads) and insert a memory barrier (`opal_atomic_wmb()` should do) in the signalling code before setting `m_signaled` then the tail of the function could look like this:

```C
    while (1) {
        qthread_yield();
        /* Check if someone woke me up. */
        if (waiter.m_signaled) {
            break;
        }
    }
    opal_mutex_lock(lock);
    return OPAL_SUCCESS;
```

Yes, the signalling will become a bit more expensive due to the memory barrier but having several threads hammering on atomic locks is by far more problematic than a few more cycles spent by one thread in the signalling code.

File: opal/mca/threads/qthreads/threads_qthreads_module.c
Comment (devreal) [reply opal/mca/threads/qthreads/threads_qthreads_module.c:88]: What is the cost of qthread_initialize()? I wonder if it's worth calling it on each and every operation. Sooner or later qhtreads will be initialized if this module is used so why not make sure it is properly initialized from the beginning

File: opal/mca/threads/qthreads/threads_qthreads_mutex.c
Comment (olivier-snl) [reply opal/mca/threads/qthreads/threads_qthreads_mutex.c:96]: Indeed this type of polling is probably not the best use of Qthreads.  Using FEBs instead would make for an event-driven approach.  Not sure if its best to leave it for now and reimplement later or go ahead and fix now. @janciesko

File: opal/mca/threads/qthreads/threads_qthreads_module.c
Comment (olivier-snl) [reply opal/mca/threads/qthreads/threads_qthreads_module.c:88]: Reinitializing Qthreads on every operation is not cheap and it would basically result in not being able to take advantage of the benefits of using a user-level threading library rather than pthreads.

Comment (shintaro-iwasaki) [reply opal/mca/threads/qthreads/threads_qthreads_module.c:88]: In any case, since there's no good timing to initialize Qthreads (specifically, it is hard to forbid developers from using mutex before initializing Qthreads and after finalizing Qthreads), this check is necessary. (MPICH took another approach that is ideal for ULTs; as far as I remember, all the components in MPICH must be aware of from when to when mutex/TLS can be used, but it needs tremendous engineering efforts to maintain.)

In terms of cost, I believe `qthread_initialize()` will take this `redundant call` path, so it should not be very expensive (of course, calling an external function needs some cycles).
https://github.com/Qthreads/qthreads/blob/master/src/qthread.c#L814-L816

Maybe I miss something, however. If this function calling overhead would kill the benefits of using user-level threads, I can introduce another flag to check if Qthread has been already initialized.

Comment (olivier-snl) [reply opal/mca/threads/qthreads/threads_qthreads_module.c:88]: Yes, I forgot that we have that check for already being initialized.

**Comment by shintaro-iwasaki** [comment]
I have updated the commits to address problems pointed out by @devreal and @jsquyres. Could someone please review this PR again?

**Review by devreal — APPROVED** [review]
I think the PR is OK. The performance issues in the conditional variable implementation is unresolved but that is not a hill I will die on.

**Comment by shintaro-iwasaki** [comment]
@jsquyres Could you please take a look at this PR? Is the issue of `TPKG_*` and `THREAD_*` blocking this PR?

File: opal/mca/threads/qthreads/configure.m4
Comment (jsquyres) [reply opal/mca/threads/qthreads/configure.m4]: > This implementation is basically a copy from #7675.

Understood.  I guess this was missed in that PR.  ☹️ 

I guess my question still stands: if `TPKG_*` and `THREAD_*` are being used by the framework (and higher -- all the way up in the top-level `configure.ac`), why is it up to the component to `AC_SUBST` them?  That seems weird.

Can you add a commit on there to remove the `AC_SUBSTs` of these values from the components, and instead, do them in `opal/mca/threads/configure.m4`?

**Comment by olivier-snl** [comment]
> I think the PR is OK. The performance issues in the conditional variable implementation is unresolved but that is not a hill I will die on.

The Qthreads team can spend some time working out a better performing solution in the near future.  The changes by Shintaro at least give us something correct for now.

File: opal/mca/threads/qthreads/configure.m4
Comment (shintaro-iwasaki) [reply opal/mca/threads/qthreads/configure.m4]: I added a commit, which fixed the issue (I also rebased my branch to the master):
- `TPKG_*` are not `AC_SUBSTs`; they appear only in `opal/mca/threads/<THREAD_PKG_NAME>/configure.m4`.
- `THREAD_*` are `AC_SUBSTs`. in `opal/mca/threads/configure.m4`.

As far as I checked, there is no compilation issue with Pthreads, Qthreads, and Argobots.

**Review by jsquyres — APPROVED** [review]

**Comment by shintaro-iwasaki** [comment]
I'd like to thank @devreal and @jsquyres for reviewing this PR! Could someone merge this PR?

Since this incomplete MCA/thread/Qthread has been in the main branch of Open MPI, we'd be extremely happy if this PR is merged before the upcoming 5.0 release. (Basically this PR does not affect the existing Pthreads implementation.)
If this PR should be rebased, more tested, or checked by additional reviewers, please let me know.

**Comment by jsquyres** [comment]
Done!  Thanks for all the work on this!

**Pull request merged**
